Client/Frontend/Browser/BrowserLongPressGestureRecognizer.swift

```diff
@@ -0,0 +1,135 @@
+
+            evaluateTouchLocation(touchLocation, gestureRecognizer: gestureRecognizer)
+        }
+    }
+
+    /// Check whether long pressed is on valid link. If found valid link then find the rect for link.
+    private func evaluateTouchLocation(touchLocation: CGPoint,gestureRecognizer: UILongPressGestureRecognizer) {
+        self.browser.webView.evaluateJavaScript("findElementsAtPoint(\(touchLocation.x),\(touchLocation.y))", completionHandler:nil)
+    }
+
+    /// Recursively call block on view and its subviews
+    private func recursiveBlockOnViewAndSubviews(mainView: UIView, block:(view: UIView) -> Void) {
+        block(view: mainView)
+        mainView.subviews.map(){ self.recursiveBlockOnViewAndSubviews($0 as UIView, block) }
+    }
+
+    /// Find location in screen corresponding to webview - in case it is zoomed or scrolled
+    private func rectLocationInWebView(webView:WKWebView,locationRect:CGRect) -> CGRect {
+        var rect = locationRect
+        var scale = self.browser.webView.scrollView.zoomScale
+        rect.origin.x *= scale
+        rect.origin.y *= scale
+        rect.size.width *= scale
+        rect.size.height *= scale
+        rect.origin.x += self.browser.webView.scrollView.contentInset.left;
+        rect.origin.y += self.browser.webView.scrollView.contentInset.top;
+
+        return rect
+    }
+
+    // MARK: - BrowserHelper Mehods
+    class func name() -> String {
+        return "BrowserLongPressGestureRecognizer"
+    }
+
+    func scriptMessageHandlerName() -> String? {
+        return "longPressMessageHandler"
+    }
+
+    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        var elementsDict: [String:AnyObject]? = message.body as? [String:AnyObject]
+        if elementsDict  == nil {
+            return
+        }
+
+        var elements:[ElementType] = []
+        if let hrefElement = elementsDict!["hrefElement"] as? [String:String] {
+            if let hrefStr: String = hrefElement["hrefLink"] {
+                if let hrefLinkURL = NSURL(string: hrefStr) {
+                    elements.append(ElementType.hrefLink(hrefLinkURL))
+                }
+            }
+        }
+
+        if let imageElement = elementsDict!["imageElement"] as? [String:String] {
+            if let imageSrcStr: String = imageElement["imageSrc"] {
+                if let imageSrc = NSURL(string: imageSrcStr) {
+                    elements.append(ElementType.ImageSrc(imageSrc))
+                }
+            }
+        }
+        var disableGestures:[UIGestureRecognizer] = []
+        self.recursiveBlockOnViewAndSubviews(self.browser.webView) { view in
+            if let gestureRecognizers = view.gestureRecognizers as? [UIGestureRecognizer] {
+                for g in gestureRecognizers {
+                    if g != self  && g.enabled == true {
+                        g.enabled = false
```

Comment:
Note that disabling all gesture recognizers unconditionally will break the default gesture recognizer for other long press actions, namely text selection. Thankfully, the fix for this should be simple: just move this chunk of code (along with the <code>disableGestures.map(...)</code> that follows) into the if condition below. That way, if there are context menu elements, we show the context menu; otherwise, we still allow the default text selection handler to kick in.